text/0000-binary-type.md

```diff
@@ -0,0 +1,86 @@
+
+## The `binary` type
+The easiest way to think about it is that the `binary` type represents an array of bytes.
+
+## Serialization
+
+This is codec dependant and not always symmetric, can not represent binary data and is a general-purpose codec it will default to base64 encoding the data as a string (JSON for example). This is an asymmetric change as we will not by default decode a base64 string as binary given without context we can't determine if this is the desired behavior or not, this can however be done using tremor-scripts functions.
+
+## Syntax
+
+We use the (erlang bit syntax)[https://erlang.org/doc/programming_examples/bit_syntax.html] as an inspiration. Tremor script gains support for binary semi-literals using the form of `<< expr1, expr2 >>` where `expr` has to return a 1-byte number. More elaborate forms, such as bitwise size and different types are aimed to be added later and follow the erlang bit syntax.
+
+Binary matching will likely follow the same logical pattern as other matching syntaxes, and aim to provide at least erlang style dissection of binary data.
+
+## stdlib
+
+This RFC introduces the `base64` module for encoding and decoding `binary` data as base64 strings. It also adds the `binary` module with basic functions such as `len`, as well as conversion functions to and from arrays of numbers.
+
+The `string` module gains functions to convert a string to a `binary` as well as a function that will take a `binary` and convert it to a utf8 `string` in a potentially lossy fashion for invalid utf8 data.
+
+The `type` module will gain the `is_binary` function as well as `type::as_string` now returning `"binary"` when appropriate.
+
+## codecs
+
+This RFC introduces the `binary` codec that passes bytes unmodified as binary data.
+
+The `json` codec is changed to serialize `binary` data as a base64 encoded string, however it will not automatically decode base64 strings as binary.
+
+The `msgpack` codec should encode and decode binary as `bin` types.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `binary` type is a `Vec<u8>` that serializes as base64 encoded string in simd-json and as `bytes`.
+
+It the new implemented functions are all constant, so will be able to be pre-computed when constant folding is possible.
+
+The binary semi-literals will, whenever possible be turned into full-literals using constant folding, but remain constructs where not following the example of arrays and records.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This breaks symetry with the JSON representation, however all differences can be recovered using the provided tremor-script functionality.
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+A quick search showed no decent alternative binary syntax to erlangs binary forms, however any other suggestions are welcome.
```

Comment:
python 3 has a  bytes literal: https://docs.python.org/3/reference/lexical_analysis.html#grammar-token-bytesprefix

Reply:
But that only looks more like strings with escape sequences or do I misread that?

Reply:
Literals look like strings but seem to represent byte sequences:

> While bytes literals and representations are based on ASCII text, bytes objects actually behave like immutable sequences of integers

https://docs.python.org/3/library/stdtypes.html#bytes

Reply:
I still wouldn't really file that under "binary literals" they are pretty much the same binary literal as `"\0x01\0x02\0x03"` in C would be or an array of integers for that matter.

Reply:
It might not be as sophisticated as erlangs syntax, but its something and is explicitly meant for encoding byte arrays, which also have their own type.